bootstore/src/schemes/v0/peer.rs

```diff
@@ -0,0 +1,802 @@
+                        .iter()
+                        .map(|(id, handle)| (id.clone(), handle.addr))
+                        .collect(),
+                    accepted_connections: self
+                        .accepted_connections
+                        .keys()
+                        .cloned()
+                        .collect(),
+                    negotiating_connections: self
+                        .negotiating_connections
+                        .keys()
+                        .cloned()
+                        .collect(),
+                };
+                let _ = responder.send(status);
+            }
+            PeerApiRequest::Shutdown => {
+                self.shutdown = true;
+                // Shutdown all connection processing tasks
+                for (_, handle) in &self.accepted_connections {
+                    let _ = handle.tx.send(MainToConnMsg::Close).await;
+                }
+                for (_, handle) in &self.negotiating_connections {
+                    let _ = handle.tx.send(MainToConnMsg::Close).await;
+                }
+                for (_, handle) in &self.connections {
+                    let _ = handle.tx.send(MainToConnMsg::Close).await;
+                }
+            }
+        }
+    }
+
+    // Route messages to their destination connections
+    async fn deliver_envelopes(&mut self) {
+        for envelope in self.fsm.drain_envelopes() {
+            if let Some(conn_handle) = self.connections.get(&envelope.to) {
+                debug!(
+                    self.log,
+                    "Sending {:?} to {}", envelope.msg, envelope.to
+                );
+                if let Err(e) = conn_handle
+                    .tx
+                    .send(MainToConnMsg::Msg(Msg::Fsm(envelope.msg)))
+                    .await
+                {
+                    warn!(self.log, "Failed to send {e:?}");
+                }
+            } else {
+                warn!(self.log, "Missing connection to {}", envelope.to);
+            }
+        }
+    }
+
+    // Perform any operations required by a given `ApiOutput`, such as
+    // persisting state, and then inform any callers (via outstanding responders)
+    // of the result.
+    async fn handle_api_output(&mut self, output: ApiOutput) {
+        info!(self.log, "Fsm output = {output:?}");
+        match output {
+            // Initialization is mutually exclusive
+            ApiOutput::PeerInitialized | ApiOutput::RackInitComplete => {
+                if let Some(responder) = self.init_responder.take() {
+                    let _ = responder.send(Ok(()));
+                }
+                // TODO: Persistence
+            }
+            ApiOutput::RackSecret { secret, .. } => {
+                // We only allow one outstanding request currently, so no
+                // need to get the `request_id` from destructuring above
+                if let Some(responder) = self.rack_secret_responder.take() {
```

Comment:
Good call! The original FSM code actually returned `ApiOutput::RackSecret` on learn requests completing, but that doesn't happen anymore, as there was no pending api call asking for the rack secret. The rack secret was just loaded as a side effect so that the `LearnedSharePkg` to distribute could be be created with a decrypted share. The decryption key is derived from the rack secret.

Reply:
Done in f354ee9c5